src/skill.cpp

```diff
@@ -225,50 +257,101 @@ void SkillLevel::train( int amount, bool skip_scaling )
     }
+
+    if( _rustAccumulator < 0 ) {
+        _rustAccumulator = 0;
+    }
+    if( _level == _knowledgeLevel && _exercise > _knowledgeExperience ) {
+        _knowledgeExperience = _exercise;
+    }
+
+    if( _knowledgeExperience >= 10000 * ( _knowledgeLevel + 1 ) * ( _knowledgeLevel + 1 ) ) {
+        _knowledgeExperience = 0;
+        ++_knowledgeLevel;
+    }
 }
 
-namespace
-{
-time_duration rustRate( int level )
+
+void SkillLevel::knowledge_train( int amount, int npc_knowledge, bool skip_scaling )
 {
-    // for n = [0, 7]
-    //
-    // 2^18
-    // -------
-    // 2^(18-n)
+    float level_gap = 1.0f;
+    // when your _level is the same or 1 level below your knowledge, gain xp at the normal rate.
+    // as your practical skill lags behind your knowledge, it gets harder to contextualize that
+    // theoretical knowledge, and your ability to learn the theory gets slower.
+
+    // The same formula applies to NPCs teaching you, but in that case the level decreases as their knowledge
+    // level exceeds your own.  The best teacher is one who is only somewhat more knowledgeable than you.
+    if( npc_knowledge > 0 ) {
+        // This should later be modified by NPC teaching proficiencies.
+        level_gap = std::max( npc_knowledge * 1.0f - _knowledgeLevel * 1.0f, 1.0f );
+    } else {
+        // Some day this should be affected by json specific to the skill, some skills are more amenable
+        // to book learning.
+        level_gap = std::max( _knowledgeLevel * 1.0f - _level * 1.0f, 1.0f );
+    }
+    float level_mult = 2.0f / ( level_gap + 1.0f );
+    amount *= level_mult;
+
+    if( skip_scaling ) {
+        _knowledgeExperience += amount;
+    } else {
+        const double scaling = get_option<float>( "SKILL_TRAINING_SPEED" );
+        if( scaling > 0.0 ) {
+            _knowledgeExperience += std::ceil( amount * scaling );
+        }
+    }
+
+    if( _knowledgeExperience >= 10000 * ( _knowledgeLevel + 1 ) * ( _knowledgeLevel + 1 ) ) {
+        _knowledgeExperience = 0;
+        ++_knowledgeLevel;
+    }
 
-    unsigned const n = clamp( level, 0, 7 );
-    return time_duration::from_turns( 1 << ( 18 - n ) );
 }
-} //namespace
 
 bool SkillLevel::isRusting() const
 {
     return get_option<std::string>( "SKILL_RUST" ) != "off" && ( _level > 0 ) &&
-           calendar::turn - _lastPracticed > rustRate( _level );
+           _rustAccumulator > 0;
 }
 
-bool SkillLevel::rust( int rust_resist, int character_rate )
+bool SkillLevel::rust( int rust_resist )
 {
-    const time_duration delta = calendar::turn - _lastPracticed;
-    const float char_rate = character_rate / 100.0f;
-    const time_duration skill_rate = rustRate( _level );
-    if( to_turns<int>( skill_rate ) * char_rate <= 0 || delta <= 0_turns ||
-        delta % ( skill_rate * char_rate ) != 0_turns ) {
+
+    if( _level >= MAX_SKILL ) {
+        // don't rust any more once you hit the level cap, at least until we have a way to "pause" rust for a while.
```

Comment:
I think a fairly straightforward solution is to allow exercise to progess up to X% of the way to MAX_LEVEL + 1, but not display this progress or allow MAX_LEVEL + 1 to be reached, then there's a whole level worth of exercise you can gain to buffer losing that last level.